proposals/format.md

```diff
@@ -0,0 +1,320 @@
+- Allowing for interpolated `string` to bind to more efficient `string.Format` overloads.
+
+### Extending params
+The language will allow for `params` in a method signature to have the types `Span<T>`, `ReadOnlySpan<T>` and 
+`IEnumerable<T>`. The same rules for invocation will apply to these new types that apply to `params T[]`:
+
+- Can't overload where the only difference is a `params` keyword.
+- Can invoke by passing a series of arguments that are implicitly convertible to `T` or a single `Span<T>` / 
+`ReadOnlySpan<T>` / `IEnumerable<T>` argument.
+- Must be the last parameter in a method signature.
+- Etc ... 
+
+The `Span<T>` and `ReadOnlySpa<T>` variants will be referred to as `Span<T>` below for simplicity. In cases where the 
+behavior of `ReadOnlySpan<T>` differs it will be explicitly called out. 
+
+The advantage the `Span<T>` variants of `params` provides is it gives the compiler great flexbility in how it allocates
+the backing storage for the `Span<T>` value. With a `params T[]` the compiler must allocate a new `T[]` for every 
+invocation of a `params` method. Re-use is not possible because it must assume the callee stored and reused the 
+parameter. This can lead to a large inefficiency in methods with lots of `params` invocations.
+
+Given `Span<T>` variants are `ref struct` the callee cannot store the argument. Hence the compiler can optimize the 
+call sites by taking actions like re-using the argument. This can make repeated invocations very efficient as compared
+to `T[]`. The langauge though will make no specific guarantees about how such callsites are optimized. Only note that 
+the compiler is free to use values other than `T[]` when invoking a `params Span<T>` method. 
+
+One such potential implementation is the following. Consider all `params` invocation in a method body. The compiler 
+could allocate an array which has a size equal to the largest `params` invocation and use that for all of the 
+invocations by creating appropriately sized `Span<T>` instances over the array. For example:
+
+``` csharp
+static class OneAllocation {
+    static void Use(params Span<string> spans) {
+        ...
+    }
+
+    static void Go() {
+        Use("jaredpar");
+        Use("hello", "world");
+        Use("a", "longer", "set");
+    }
+}
+```
+
+The compiler could choose to emit the body of `Go` as follows:
+
+``` csharp
+    static void Go() {
+        var args = new string[3];
+        args[0] = "jaredpar";
+        Use(new Span<int>(args, start: 0, length: 1));
+
+        args[0] = "hello";
+        args[1] = "world";
+        Use(new Span<int>(args, start: 0, length: 2));
+
+        args[0] = "a";
+        args[1] = "longer";
+        args[2] = "set";
+        Use(new Span<int>(args, start: 0, length: 3));
+   }
+```
+
+This can siginficantly reduce the number of arrays allocated in an application. Allocations can be even further 
+reduced if the runtime provides utilities for smarter stack allocation of arrays.
```

Comment:
could you expand on this further?

Reply:
The later sections go into a few of the possible scenarios here around stack allocation.